stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -1,80 +1,141 @@
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     * @param x Source matrix
+     * @param y Target matrix
      */
-    inline std::vector<int>
-    append_array(const std::vector<int>& x, const std::vector<int>& y) {
-      std::vector<int> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
-      return z;
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_copy(const Eigen::Matrix<T1, R, C>& x,
+                     Eigen::Matrix<T2, R, C>& y) {
+      y = x.template cast<T2>();
     }
 
     /**
-     * Return the concatenation of two vectors containing Matrices
-     *   of the same type in the order of the arguments.
+     * This function copies and casts the elements of std::vector x to std::vector y.
+     * First, y is resized to match the size of x, and then append_array_copy
+     * is recursively called on each element of x to copy it to an element of y.
      *
-     * @tparam T1 Type of element in Matrices in first vector
-     * @tparam T2 Type of element in Matrices in second vector
-     * @tparam R Row specification of matrices
-     * @tparam C Column specification of matrices
-     * @param x First vector
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * @tparam Type of source
+     * @tparam Type of destination
+     * @param x Source variable
+     * @param y Target variable
      */
-    template <typename T1, typename T2, int R, int C>
-    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
-    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
-                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    template<typename T1, typename T2>
+    inline void append_array_copy(const std::vector<T1> &x,
+                                  std::vector<T2> &y) {
+      y.resize(x.size());
+      for (size_t i = 0; i < x.size(); i++) {
+        append_array_copy(x[i], y[i]);
       }
+    }
 
-      std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> > z;
-      z.reserve(x.size() + y.size());
-      for (size_t i = 0; i < x.size(); i++)
-        z.push_back(x[i].template cast<typename return_type<T1, T2>::type>());
-      for (size_t i = 0; i < y.size(); i++)
-        z.push_back(y[i].template cast<typename return_type<T1, T2>::type>());
-      return z;
+    /**
+     * This function checks the sizes of both arguments are consistent.
+     *
+     * The basic version assumes variables are of compatible sizes. There are
+     * specializations for Eigen::Matrices and std::vectors (where this
```

Comment:
There is no Eigen::Matrices, it's Eigen::Matrix.  I also didn't understand this comment.  If the variables aren't `Eigen::Matrix or `std::vector` types, then they're scalars and have to be the same size.   Did you just mean "scalar" instead of "compatible sizes"?

What happens if you pass in one scalar and one vector?  Or is that never going to happen because of the outside calling structure?  If so, you might want to say that there's an underlying assumption the two arguments are of the same shape, if not the same scalar type.

There's no notion of "basic version".  This can be a base template function (which can get specialized) or just a standalone overload.